### ISSUE TITLE
[Discussion] What are our norms as a community around development on branches vs forks? 

### ISSUE THREAD

**Issue by Arielle-Bennett** [the-turing-way/the-turing-way#3871]
### Summary

This question came up at the Book Dash Intro to GitHub session on 28 October 2024: should we be asking people to make their own forks of the project and make edits there OR develop on branches within the turing-way/turing-way repo. 

Both have pros and cons, including reflecting other open source norms (forks) to managing build previews with our current infrastructure (branches?) 

I'd like to have a discussion on what the community (and in particular maintainers) views are on this and whether we want to establish community practices around the decision.

### What needs to be done?

- [ ] Discuss what maintainers would like the community norms to be
- [ ] Make an initial agreement on community norms 
- [ ] Update documentation and materials to reflect the decision
- [ ] Teach new members the preferred approach going forward

### Who can help?

* Maintainers and community members, please weigh in with your thoughts!

---

### Updates

*Lorem ipsum dolor sit amet, consectetur adipiscing.*

**Comment by EstherPlomp** [comment]
I use branches because it is easier to quickly edit an existing page + building previews is less complicated I assume? I'm not super technical but can learn if we're going for forking I suppose..

**Comment by KirstieJane** [comment]
I think if we _don't_ demonstrate forking then we're not meeting a goal of setting up Turing Way contributors to be able to contribute to other open source projects....

So I personally think nudging towards forks is better for folks in the long run... but if there are things that can't be done from a fork (ghpages preview?) then we need to grapple with how much we're willing to make the first few contributions hard!!

I think a stronger (and more contentious!) opinion is that we need to run regular trainings that are a level up from the intro to GitHub where we more explicitly cover merge conflicts and working locally... and extend that to a (maybe unenforceable??) requirement that maintainers know the best practices for contributions to open source projects??

**Comment by JimMadge** [comment]
I tend to push my branches to this repo, because I can, which might not be a great reason.

> I think if we don't demonstrate forking then we're not meeting a goal of setting up Turing Way contributors to be able to contribute to other open source projects....

I think @KirstieJane makes a good point. Also, if we (the core team) don't use forks we are making a division between us and the rest of the community. In some projects that makes sense but I don't think that is the culture we want to create here.

Related to this, enabling all contributors to push their branches to this repo, and merge their own PRs, means we have to add them as collaborators with write permissions. I think that is probably more permission than you want to give everyone. I think there was a hope migrating the repository would be a good place to change that. At the moment, there are 233 accounts with write or greater permissions who are not members of the org. There is no team for those accounts. I would guess that is already unmanagable.

> So I personally think nudging towards forks is better for folks in the long run... but if there are things that can't be done from a fork (ghpages preview?) then we need to grapple with how much we're willing to make the first few contributions hard!!

Running certain actions from forks may not be possible, or might require approval. I wouldn't let that dictate the contribution workflow though.

> I think a stronger (and more contentious!) opinion is that we need to run regular trainings that are a level up from the intro to GitHub where we more explicitly cover merge conflicts and working locally... and extend that to a (maybe unenforceable??) requirement that maintainers know the best practices for contributions to open source projects??

I like that idea 🚀! I think the processes can be inconsistent, depending on who your reviewer is. It can cause problems as bugs slip by. Something like triage/reviewer training and guidelines could help.

**Comment by KirstieJane** [comment]
From @JimMadge 

> Related to this, enabling all contributors to push their branches to this repo, and merge their own PRs, means we have to add them as collaborators with write permissions. I think that is probably more permission than you want to give everyone. I think there was a hope migrating the repository would be a good place to change that. At the moment, there are 233 accounts with write or greater permissions who are not members of the org. There is no team for those accounts. I would guess that is already unmanagable.

So - I think this is a good time to reflect and review BUT I will say that to date the adding contributors so they can merge their own PRs and so they _ideally_ can be reviewers very _easily_ (without asking for explicit permission) has been by design.

I don't see many bugs slipping through the CIs and the original philosophy was that it should be safe to fail within The Turing Way (and most importantly it's safe to TRY!) but I'm open to feedback that it's too hard to manage at scale.

For me though it's the role modelling that I think we should focus on, keeping the permissions pretty permissive!

**Comment by JimMadge** [comment]
>  adding contributors so they can merge their own PRs and so they ideally can be reviewers very easily (without asking for explicit permission) has been by design.

Yes absolutely. I didn't mean to suggest anyone was doing anything wrong. I remember when we had conversations at repo migration time it was something people strongly wanted to keep.

> I don't see many bugs slipping through the CIs and the original philosophy was that it should be safe to fail within The Turing Way (and most importantly it's safe to TRY!) but I'm open to feedback that it's too hard to manage at scale.

It might be more difficult than you think to manage. Sometimes I see things which I feel I would have wanted to address before merging. However, maybe that is me applying a different standard 😅.

We have gotten better over time, introducing more CI. However, that is a balancing act of trying to maintain quality while not introducing too many barriers. One example is the recent addition of validation for the all contributors JSON file, which often had problems before and would cause the table to render incorrectly. Another might be Markdown, which I don't think we lint at all. So, it is up to reviewers to check the style and rendering as anything which can build will pass CI.

I think it is a difficult balance, so I'm not really saying the current state is wrong or another state is categorically better.

It might be cool to explore tools like pre-commit bot and pettier, where a bot could enforce a style in PRs. That way you get strict style without passing that burden onto either the contributors or reviewers :thinking:.

**Comment by KirstieJane** [comment]
@JimMadge - all great points to discuss! Keen to hear others' preferences 😀

**Comment by EKaroune** [comment]
FYI - have not had time to read all detail of above, so sorry if any of this repeats.
I think branching is better option than forking as keeps contributions in our repository so they are more visible and accessible to community to comment and contribute to. As Jim has said above, it is often hard to contribute to other peoples forks as there are access issues.  
Also from technical point of view - branching is much easier for GH beginners and people often get caught up bringing forks into our repo, more conflicts, etc.
I do get your point though @KirstieJane about it being part of open source working, so we probably need to provide that extra more intermediate GH training.

**Comment by KirstieJane** [comment]
Totally see your points @EKaroune! Thanks for commenting!

**Comment by JimMadge** [comment]
Some good points about forks from @EKaroune.

I don't think forks are any less visible or accessible as long as people open PRs (and allow maintainers to edit their PRs, which I _think_ is default). I doubt people go looking through the branches without PRs to found out what is going on.

Speaking of which, we have quite a few stale branches. Encouraging people to use forks would help clean that up too.

Probably the main challenge with encouraging people to use forks is ensuring that they keep their fork up to date with this repo (otherwise difficult merge conflicts are likely). We could make that part of the training though, if it isn't already.

I think the best way is to use two remotes, like

```console
$ git clone git@github.com/JimMadge/the-turing-way.git
$ git remote add upstream https://github.com/the-turing-way/the-turing-way.git
```

Then before starting any new work

```console
$ git switch main
$ git fetch upstream
$ git merge --ff-only upstream/main main
```

or

```console
$ git switch main
$ git pull --ff-only upstream main
```

I also think that although the forking workflow is a little more complicated, it is probably a better way to teach new contributors. First, they don't have to be added as a contributor to push to their own fork (whereas they will to push branches to this repo). Second, I think it actually gives a much better impression of how git works. It isn't a client-server VCS where there is a canonical repository, every clone (including forks and local clones) are independent, complete repositories. When you want to merge a branch from one repository into another, you need to resolve any conflicts.

**Comment by Arielle-Bennett** [comment]
Thanks @JimMadge for your comments on this - I think it's really useful and I definitely take @EKaroune / @EstherPlomp points about branches being more accessible for people totally new to Git / GitHub or those with less technical knowledge (I know it's easier for me :D ) 

I think we could be explicit about supporting both methods, but I do think we should be teaching /training people to fork if we can, since it's how a lot of other open source projects work and we want them to build skills here that they can take onto other projects. It's okay if people have to deal with merge conflicts etc because this is a good environment to learn that anyway and we now have a handy button for syncing up forks so they don't cause issues when someone wants to do a pull request...

**Comment by JimMadge** [comment]
Yeah, I think the education argument for forking is strong.

It is very unlikely you would be added as a contributor to project until you have made a number of contributions or joined the core team.

**Comment by RichardJActon** [comment]
I checked the community handbook for what it currently has to say on this question and it only mentions the forking approach, so currently the documentation very much points people towards that way of working:

https://deploy-preview-3940--the-turing-way.netlify.app/community-handbook/contributing-guideline#making-a-change-with-a-pull-request

I favor the forking approach for education purposes and in the interests of keeping the main repo tidy. It's how I made my first contribution to TTW, as I had not been added as a contributor to the repo when I first contributed. (I was a little surprised when I was added shortly afterwards, as, in my experience github contributors where mostly small numbers of trusted project insiders.

If we do decide to go down this route we'll need to evacuate
